flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/AbstractHandler.java

```diff
@@ -117,13 +117,10 @@ protected void respondAsLeader(ChannelHandlerContext ctx, RoutedRequest routedRe
 
 		FileUploads uploadedFiles = null;
 		try {
-			synchronized (this) {
-				if (terminationFuture != null) {
-					log.debug("The handler instance for {} had already been closed.", untypedResponseMessageHeaders.getTargetRestEndpointURL());
-					ctx.channel().close();
-					return;
-				}
-				inFlightRequestTracker.registerRequest();
+			if (!inFlightRequestTracker.registerRequest()) {
```

Comment:
So this works because the final unregister happens when the closeHandlerAsync future completes, which does so once the async operation was accessed from the `completedOperationsCache`?

Reply:
No, this works because we terminate the `AbstractHandler` (completing the `terminationFuture`) once all in flight requests are completed. See this line: untime/src/main/java/org/apache/flink/runtime/rest/handler/AbstractHandler.java#L253. Afterwards the `inFlightRequestTracker` will no longer accept new requests because it has already terminated.

What won't work properly though is if a new request comes which creates a new asynchronous operation. I think for this to work, the `CompletedOperationCache` needs to track whether `closeAsync` had been called or not.